app/src/main/java/api/RetrofitAPI.kt

```diff
@@ -8,14 +8,13 @@ import java.net.UnknownHostException
 // Note: Fetching users-list is the *only* use-case therefore avoid implementing unrelated,
 // extra, features no matter how common or useful they might be for potential future cases.
 //
-class RetrofitAPI() : API {
+class RetrofitAPI(private val usersListURL: String) : API {
 
     /**
      * Declare and Initialize user api service lazily whenever required
      */
-    private val userApiService by lazy {
-        API.create()
-    }
+    private val retrofit by lazy { NetworkModule.provideRetrofit(usersListURL) }
+    private val userApiService by lazy { NetworkModule.provideUserApi(retrofit) }
 
     override fun fetchUsersList(
```

Comment:
- This function executes network fetch inside `runBlocking`. What does run-blocking mean and why is it the right choice here? 
- Upon receiving response from server, `failure` with `IllegalArgumentException` is invoked when no data is received. Is this the right exception to throw? (Refer to IllegalArgumentException documentation)

Reply:
@nets-dev

Q1] This function executes network fetch inside runBlocking. What does run-blocking mean and why is it the right choice here?
**Ans** :-  As the network call is async, hence we have to use await() to get the API result. As await() is a suspended function, it has to handle inside coroutine i.e run-blocking/launch. run-blocking is a coroutine builder which waits for completion. 
    
Q2] Upon receiving response from server, failure with IllegalArgumentException is invoked when no data is received. Is this the right exception to throw? (Refer to IllegalArgumentException documentation).
**Ans** :- IllegalArgumentException is used because whenever we request a server with an illegal parameter we are getting a bad request or null error. Here in this case if I am getting null or bad requests it should specifically mention error due to Illegal arguments.